jobapi/routes.go

```diff
@@ -0,0 +1,147 @@
+package jobapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/buildkite/agent/v3/agent"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	"golang.org/x/exp/maps"
+)
+
+// router returns a chi router with the jobapi routes and appropriate middlewares mounted
+func (s *Server) router() chi.Router {
+	r := chi.NewRouter()
+	r.Use(
+		middleware.Recoverer,
+		// middleware.Logger, // REVIEW: Should we log requests to this API? If so, where should we log them to? The job logs?
+		HeadersMiddleware(),
+		AuthMiddleware(s.token),
+	)
+
+	r.Route("/api/current-job/v0", func(r chi.Router) {
+		r.Get("/env", s.getEnv)
+		r.Patch("/env", s.patchEnv)
+		r.Delete("/env", s.deleteEnv)
+	})
+
+	return r
+}
+
+func (s *Server) getEnv(w http.ResponseWriter, _ *http.Request) {
+	normalizedEnv := s.environ.Dump()
+	resp := EnvGetResponse{Env: normalizedEnv}
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(resp)
+}
+
+func (s *Server) patchEnv(w http.ResponseWriter, r *http.Request) {
+	var req EnvUpdateRequest
+	err := json.NewDecoder(r.Body).Decode(&req)
+	defer r.Body.Close()
+	if err != nil {
+		writeError(w, fmt.Errorf("failed to decode request body: %w", err), http.StatusBadRequest)
+		return
+	}
+
+	added := make([]string, 0, len(req.Env))
+	updated := make([]string, 0, len(req.Env))
+	protected := checkProtected(maps.Keys(req.Env))
+
+	if len(protected) > 0 {
+		writeError(
+			w,
+			fmt.Sprintf("the following environment variables are protected, and cannot be modified: % v", protected),
+			http.StatusUnprocessableEntity,
+		)
+		return
+	}
+
+	nils := make([]string, 0, len(req.Env))
+
+	for k, v := range req.Env {
+		if v == nil {
+			nils = append(nils, k)
+		}
+	}
+
+	if len(nils) > 0 {
+		writeError(
+			w,
+			fmt.Sprintf("removing environment variables (ie setting them to null) is not permitted on this endpoint. The following keys were set to null: % v", nils),
+			http.StatusUnprocessableEntity,
+		)
+		return
+	}
+
+	for k, v := range req.Env {
+		if _, ok := s.environ.Get(k); ok {
+			updated = append(updated, k)
+		} else {
+			added = append(added, k)
+		}
+		s.environ.Set(k, *v)
+	}
+
+	resp := EnvUpdateResponse{
+		Added:   added,
+		Updated: updated,
+	}
+
+	resp.Normalize()
+
+	json.NewEncoder(w).Encode(resp)
+	w.WriteHeader(http.StatusOK)
```

Comment:
w.WriteHeader needs to come before writing to w ... if for no other reason than it logs a warning about a superflouous WriteHeader call.